### ISSUE TITLE
How do we handle python dependencies

### ISSUE THREAD

**Issue by Makman2** [coala-analyzer/coala#1852]
This clashes inside the CI: https://circleci.com/gh/coala-analyzer/coala/5434
Debug output from my Fedora 23 machine:
```
mak@localhost:~/dev/coala (master) +1 ~0 -0 $ coala-ci -L=DEBUG
[ERROR][23:05:09] An unknown error occurred. This is a bug. We are sorry for the inconvenience. Please contact the developers for assistance. During execution of coala an exception was raised. This should never happen. When asked for, the following information may help investigating:
[DEBUG][23:05:09] Exception was:
Traceback (most recent call last):

  File "/usr/lib/python3.4/site-packages/coalib/coala_main.py", line 60, in run_coala
    autoapply=autoapply)

  File "/usr/lib/python3.4/site-packages/coalib/settings/ConfigurationGathering.py", line 255, in gather_configuration
    log_printer)

  File "/usr/lib/python3.4/site-packages/coalib/settings/SectionFilling.py", line 29, in fill_settings
    bear_dirs = section.bear_dirs()

  File "/usr/lib/python3.4/site-packages/coalib/settings/Section.py", line 73, in bear_dirs
    for bear_dir in collect_registered_bears_dirs('coalabears')]

  File "/usr/lib/python3.4/site-packages/coalib/collecting/Collectors.py", line 209, in collect_registered_bears_dirs
    registered_package = ep.load()

  File "/usr/lib/python3.4/site-packages/pkg_resources/__init__.py", line 2201, in load
    self.require(*args, **kwargs)

  File "/usr/lib/python3.4/site-packages/pkg_resources/__init__.py", line 2218, in require
    items = working_set.resolve(reqs, env, installer)

  File "/usr/lib/python3.4/site-packages/pkg_resources/__init__.py", line 835, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)

pkg_resources.ContextualVersionConflict: (colorama 0.3.7 (/usr/lib/python3.4/site-packages), Requirement.parse('colorama==0.3.6'), {'radon'})
```

**Comment by Makman2** [comment]
This is a problem inside coala-bears, gets solved with https://github.com/coala-analyzer/coala-bears/pull/186.

**Comment by AbdealiJK** [comment]
I guess we need to use >= in pyprint ?

-----Original Message-----

**Comment by AbdealiJK** [comment]
@Makman2 and I had a discussion on gitter about this.

The points we had were:
 - If we use `==` and the user has a VersionConflict the user is expected to use a new virtualenv for coala.
   - Argument against this: If all packages did this, a user will need lots of virtualenvs which is unmanageable. For example a website which uses flask, coala, pytest and numpy(for stats) would need 4 virtualenvs. Which is confusing and complicated. Hence this doesnt seem the correct way for a python package manager to do things.
 - We use `>=` and test with as many packages too !
    - Arguments against this: This is again complicated because CI times will be much more and things blow up.
 - We use `>=` and figure out better testing methods
    - The point here is that we do the whole `>=` thing but we make our tests smarter.
    - We can add something to test with latest versions. But we shouldnt trigger this for every PR as it'd be slow (no caches).
    - Maybe run some CI systems based on a trigger. This would be like we do basic tests by default for every PR. We do an extensive test only if a maintainer comments with "do all tests" or similar. Orgs like mozilla and wikimedia do this 

TL;DR : The log from IRC
```
<coala_gitter> [AbdealiJK] Can you give me proper context ? As I see it we have a version conflict with some other package because radon uses `==` - is this right ?
<coala_gitter> [Makman2] yes
<coala_gitter> [AbdealiJK] And this package conflist is colorama - which is being used in pyprint and radon
<coala_gitter> [Makman2] @sils1297 finally convinced me actually using `==` because otherwise we can't guarantee our stuff working
<coala_gitter> [Makman2] yes :3
<coala_gitter> [AbdealiJK] @Makman2 can you convince me now :D
<coala_gitter> [AbdealiJK] Cuz @sils1297 isnt online
<coala_gitter> [Makman2] main reason is we can't say coala works without testing it completely
<coala_gitter> [AbdealiJK] @Makman2 At some point we need to upgrade our deps right - so when do we do this ?
<coala_gitter> [Makman2] currently we test with the most recent versions, but as soon as radon (or some other package) changes something, we can't predict what happen^^
<coala_gitter> [Makman2] two options:
<coala_gitter> [AbdealiJK] I mean we cant use setuptools v18 when v23 is out ...
<coala_gitter> [Makman2] doing that manually, so we just make a commit with specifying higher versions, let our CI infrastructure test that, and see if we need adaptions etc.
<coala_gitter> [Makman2] or we try getting this deps.io website to work (or how it was called again)
<coala_gitter> [Makman2] that automatically sends pull requests if new versions arise
<coala_gitter> [AbdealiJK] Wow, thats cool
<coala_gitter> [AbdealiJK] Alright - I have another alternative
<coala_gitter> [Makman2] problem is, deps.io was already tried but didn't work
<coala_gitter> [Makman2] so manually is the easiest approach for now
<coala_gitter> [Makman2] @AbdealiJK sure go ahead
<coala_gitter> [AbdealiJK]  - We use `>=` and with out regular CI stuff. The versions wouldnt normally change in between builds as we do not use the `--upgrade` flag with pip. i.e. our cached versions are preferred over the pip latest version
<coala_gitter> [AbdealiJK] So, when we do a release we run circleci without cache and use the latest versions and make sure stuff works. This is basically a required step before every release
<coala_gitter> [Makman2] but still we don't test our stuff for versions above `>=x.x.x` then
<coala_gitter> [Makman2] and suddenly the release build fails
<coala_gitter> [AbdealiJK] We do that before every release
<coala_gitter> [AbdealiJK] When we tell CIs to not use caches
<coala_gitter> [Makman2] yeah sure, but then we have much work adjusting everything at once
<coala_gitter> [Makman2] I think it's no requirement to be able to run with the most recent packages
<coala_gitter> [AbdealiJK] @Makman2 using `==` and then changing the versions together is the same
<coala_gitter> [Makman2] but you want this build only for release
<coala_gitter> [AbdealiJK] @Makman2 I disagree. A package should always work with the latest package. Do you know the `archlinux` motto ?
<coala_gitter> [Makman2] this means we merge our PRs with the version exactly specified behin `>=`
<coala_gitter> [Makman2] and suddenly for release we are inconsistent and try the most recent packages
<coala_gitter> [Makman2] no not really
<coala_gitter> [Makman2] :3
<coala_gitter> [AbdealiJK] @Makman2 so when do you propose we update the versions on `==`
<coala_gitter> [AbdealiJK] We can do the same thing with `>=` ...
<coala_gitter> [Makman2] somewhen when we think enough time passes or we see something cool arised, dunno
<coala_gitter> [Makman2] from time to time someone can run a script that gets the newest pip versions, and compares them to ours^^
<coala_gitter> [AbdealiJK] @Makman2 that is really a bad system imo ... Let me try an alternative method of explaining
<coala_gitter> [Makman2] but using `>=` isn't reliable, like it is now with all PRs, we can't merge anything^^
<coala_gitter> [AbdealiJK] If I have lets say 2 python packages I use : flask and coala
<coala_gitter> [AbdealiJK] How would I go about handling the dependencies if they have a Version Conflict ?
<coala_gitter> [Makman2] use a venv^^
<coala_gitter> [AbdealiJK] You want me to use 2 venvs ?
<coala_gitter> [AbdealiJK] 1 for coala and 2 for flask ?
<coala_gitter> [Makman2] especially coala takes on much dependencies, that makes sense^^
<coala_gitter> [Makman2] why 2 for flask?
<coala_gitter> [AbdealiJK] typo :P
<coala_gitter> [Makman2] ah okay :D
<coala_gitter> [Makman2] yeah
<coala_gitter> [Makman2] one for both
<coala_gitter> [AbdealiJK] And lets say I add pytest to this - you want me to use 3 ?
<coala_gitter> [Makman2] depends on where you want to use it
<coala_gitter> [AbdealiJK] And if I have lets say scripts which use numpy/scipy for database migration and analytics i use 4 ?
<coala_gitter> [AbdealiJK] @Makman2 I use all of these for my website
<coala_gitter> [Makman2] inside the coala venv coala installs you the right version, in flask venv it installs you the needed version for flask
<coala_gitter> [AbdealiJK] I find that is just weird.
<coala_gitter> [Makman2] why that?
<coala_gitter> [AbdealiJK] You want a virtualenv for every package by that logic
<coala_gitter> [Makman2] no
<coala_gitter> [AbdealiJK] And I will spend half my dev time switching virtualenvs ...
<coala_gitter> [Makman2] if you have version conflicts there's no other choice
<coala_gitter> [Makman2] but you won't switch permanently
<coala_gitter> [AbdealiJK] @Makman2 thats the reason for using `>=` ... you should never use `==` if youre providing a package.
<coala_gitter> [Makman2] you install the stuff you need for flask inside the flask venv (since you use database migration actually there and not inside coala)
<coala_gitter> [Makman2] and the stuff for coala inside the coala venv
<coala_gitter> [Makman2] alright
<coala_gitter> [Makman2] assume a user gets now coala, like me
<coala_gitter> [Makman2] runs it
<coala_gitter> [AbdealiJK] @Makman2 lets say numpy and flaska lsi adopt the `==` standard. Im assuming here that THE way to handle deps everywhere is `==`
<coala_gitter> [Makman2] and then this happens:
<coala_gitter> [AbdealiJK] Im not specifically talking about coala here ...
<coala_gitter> [Makman2] ```
<coala_gitter> mak@localhost:~/dev/coala (master) +1 ~0 -0 $ coala
<coala_gitter> [ERROR][23:05:09] An unknown error occurred. This is a bug. We are sorry for the inconvenience. Please contact the developers for assistance. During execution of coala an exception was raised. This should never happen. When asked for, the following information may help investigating:
<coala_gitter> ```
<coala_gitter> [Makman2] yes, but the principle is, the user thinks coala sucks
<coala_gitter> [AbdealiJK] Im just trying to figure out how  a python package should handle deps in general
<coala_gitter> [Makman2] or any other program that crashes now
<coala_gitter> [AbdealiJK] @Makman2 that wouldnt happen
<coala_gitter> [Makman2] we could provide something like version ranges or even `>=`, bbut the requirement is then to test *every* package that fulfill the version specs
<coala_gitter> [AbdealiJK] pip wouldnt allow installing one of the packages because of VersionConflict
<coala_gitter> [AbdealiJK] I find what we're doing here to be making our life (the package dev's) life easier while making it more difficult for the user
<coala_gitter> [Makman2] exactly
<coala_gitter> [Makman2] then he tries it with a venv
<coala_gitter> [AbdealiJK] @Makman2 He ALREADY uses a virtualenv for flask
<coala_gitter> [Makman2] then he creates another one :P
<coala_gitter> [AbdealiJK] I always use virtualenvs are one-to-one maps for an application im making
<coala_gitter> [Makman2] yeah but the intent is not to make it more difficult for the user, we just want to guarantee everything works^^
<coala_gitter> [AbdealiJK] Having multiple defeats the purpose ...
<coala_gitter> [Makman2] I don't think having multiple venvs defeats the purpose^^
<coala_gitter> [AbdealiJK] @Makman2 I'm still not convinced and I have no other alternative way of explaining this - so, I think we're at a stalemate
<coala_gitter> [Makman2] actually I believe that's what venv is for, switching environments^^
<coala_gitter> [AbdealiJK] @Makman2 not for each task in a single application
<coala_gitter> [Makman2] I would keep `>=` or similar, but then you need to write tests that test every package^^
<coala_gitter> [AbdealiJK] I'd hate to use django or flask if I had to do that for every single website that I've developed
<coala_gitter> [Makman2] but then our build times would rise into the sky^^
<coala_gitter> [AbdealiJK] @Makman2 EXACTLY
<coala_gitter> [AbdealiJK] so we're making stuff difficult for the user because we dont want to make our work more difficult
<coala_gitter> [AbdealiJK] If youre saying the problem is our CI is slow - Your solution isnt solving the problem, it's a monkey patch
<coala_gitter> [Makman2] no that's not the problem
<coala_gitter> [AbdealiJK] @AbdealiJK wonders if he's making sense
<coala_gitter> [Makman2] it's practical only test against one package version
<coala_gitter> [Makman2] we have about 20 deps
<coala_gitter> [AbdealiJK] @Makman2 let me summarize my train of though: We **should** support more versions of other packages => we need to test with more version => we need better CI
<coala_gitter> [Makman2] do you want to start permuting them somehow?
<coala_gitter> [Makman2] not cool at all
<coala_gitter> [AbdealiJK] cool isnt a good metric here o.O
<coala_gitter> [Makman2] just wanted to say that the testing time would get too high^^
<coala_gitter> [AbdealiJK] @Makman2 then we find alternative method of testing
<coala_gitter> [AbdealiJK] Again, if testing is the problem. We fix the testing.
<coala_gitter> [Makman2] but how's the question^^
<coala_gitter> [AbdealiJK] @Makman2 i think permuting them is a bad idea
<coala_gitter> [Makman2] and until we solve this, we should use `==` so everythings works, across coala-bears, coala and for the user (that is using a venv btw :P)
<coala_gitter> [AbdealiJK] But I can think of a few method which is considerably better.
<coala_gitter> [AbdealiJK] @Makman2 if we use `==` ... we're never going to solve this as we have no motivation to :D
<coala_gitter> [Makman2] there are many possibilities to permute, that doesn't mean we check each existing combination^^
<coala_gitter> [AbdealiJK] Its the concept of "if it works, let it be B-) "
<coala_gitter> [Makman2] then where's the problem? :D
<coala_gitter> [AbdealiJK] @Makman2 again - i never said we permute
<coala_gitter> [Makman2] if it works for us and the user, no one complains, no motivation, evertything's alright :D
<coala_gitter> [AbdealiJK] @Makman2 because Ive faced the versionConflict with `==` also ...
<coala_gitter> [AbdealiJK] Thats the main reason I proposed switch to `>=`
<coala_gitter> [AbdealiJK] Hey i have to go
<coala_gitter> [AbdealiJK] be back in 10
<coala_gitter> [AbdealiJK] I think we should vote ?
<coala_gitter> [Makman2] you should open an issue again for it, problem is we need to fix our CI soon
<coala_gitter> [Makman2] even if it's a monkey patch, we need that now^^
<coala_gitter> [Makman2] if you want that still being open, I can just limit the radon version number, so just fixing radon
<coala_gitter> [Makman2] though it doesn't solve the overall problem imo that we can't rely on packages we don't test agains^^
<coala_gitter> [AbdealiJK] @Makman2 If you remember the history of this problem - this is recursive
<coala_gitter> [Makman2] yeah I know :/
<coala_gitter> [AbdealiJK] I think if youre looking for a hotfix I think fixing radon is right
<coala_gitter> [AbdealiJK] But we should fix the whole `==` vs `>=` before 0.5.0
<coala_gitter> [Makman2] how do flask etc handle such cases?
<coala_gitter> [AbdealiJK] Im wondering the same. I just checked out numpy/scipy and i cant find any deps `>_>`
<coala_gitter> [Makman2] I mean with testing packages^^
<coala_gitter> [Makman2] so in fact they just hope that it works :O
<coala_gitter> [AbdealiJK] Django would be a good case to check this on. As its bigger.
<coala_gitter> [AbdealiJK] @Makman2 I think it has no deps at all, i.e. they use custom C libs everywhere
<coala_gitter> [Makman2] phew...
<coala_gitter> [AbdealiJK] @Makman2 just mentioning here - I know that wikipedia seems to have this cool jenkins command "recheck latest versions" I remember someone using that
<coala_gitter> [Makman2] alright it seems there's still some space for discussion
<coala_gitter> [AbdealiJK] Again, if testing is the issue. I think I can come up with a few solutions for that
<coala_gitter> [Makman2] I'll publish a hotfix now, @AbdealiJK please open an issue, try to summarize our discussion from here there
<coala_gitter> [Makman2] and some possible ideas/solutions (like the jenkins command)
<coala_gitter> [Makman2] In fact I wanna too support many versions from deps, that makes stuff more flexible
<coala_gitter> [AbdealiJK] :+1: will open the issue and get the opinions of all those who're sleeping right now
```

**Comment by AbdealiJK** [comment]
After thought: We should ask the IRC folks in distutils (or setuptools?) about this.

**Comment by Makman2** [comment]
good idea :+1:

**Comment by sils1297** [comment]
see also https://github.com/rubik/radon/issues/82

**Comment by sils1297** [comment]
FWIW I tend to agree with @AbdealiJK in that things that are used as libraries or in general together with other python packages - like coala since it's inherently for software development - should rather take a more flexible approach to their requirements. We might want to fix the range to not allow major version increments but only minor and micros?

**Comment by AbdealiJK** [comment]
Apparently we can use `~=` if we want to.
And even wildcards are allowed https://www.python.org/dev/peps/pep-0440/#compatible-release like `5.*`

**Comment by sils1297** [comment]
I'm in favour of using `~=` for everything though I think we need to use `==` for the coala dependency in coala-bears because I think the dev releases might be considered comatible but if we're working on breaking changes they won't be.

**Comment by sils1297** [comment]
i.e. for releases we'd also use `~=` for the coala dep

**Comment by AbdealiJK** [comment]
I think I may have gotten a break through on this ...
https://caremad.io/2013/07/setup-vs-requirement/

What I think we need to do now is have a list of `install_requires` and `requirements.txt` independently.
The install_requires will support as many version as it can of it's packages, while `requirements.txt` will lock the versions with `==`.

Does this make sense ?
So, basically before every release we need to update and test `install_requires` (maybe manually)
But devs use `requirements.txt` - so they dont get affected by third party packages.

You may think this is repetition - but in reality, it's not. These two lists mean 2 separate things although they share the package names, they do *not* share the versions. They're used in 2 different cases and are independent.

**Comment by AbdealiJK** [comment]
A few additions:
It's alright to use `==` in requirements as developers *should* make a new virtualenv for developing on coala. And for developers, coala is the application, while everything else it installs should handle as many versions as possible, as those are the libraries that our app (coala) uses.

For other users, coala itself is a library and not the application - so, in that case we should allow as many versions as possible.

For coala-bears depending on coala, we can use `~=` because they're just modular parts of the same application (for developers coala/coala-bears).

I'm wondering if I'm making sense ...

**Comment by Makman2** [comment]
It's at least a good concept we can discuss, so separating release-package-requirements and dev requirements. This also avoids overhead in our CI, since we only test *every* package on release.

**Comment by sils1297** [comment]
I have worked a lot today and implemented ~= for everything, >= for setuptools in PyPrint, coala and coala-bears it seems to work smoothly now.

**Comment by AbdealiJK** [comment]
Why >= gitter setuptools ?
Dog something break with ~= in setuptools?

-----Original Message-----

**Comment by AbdealiJK** [comment]
Did*

**Comment by sils1297** [comment]
yes, see commit message.

**Comment by sils1297** [comment]
Alright we have pretty decent handling for this now, we use `~=` where possible, `>=` for setuptools because we don't realy have a choice. During development of coala-bears we'll probably be able to use `==0.6.0.dev1234,==0.6.0.dev99999...` so we accept both from source and the exact version we work with?

**Comment by sils1297** [comment]
This issue should be closed and a new one for coala-bears should be opened, coalas handling is well right now here IMO, @AbdealiJK can you do that?

**Comment by AbdealiJK** [comment]
Reopening based on a discussion with @ sebix on gitter.

For now, we *should* decrease the version of all our `~=` requirements based on the observation:
> Should have read pep-0440 better.~= 0.3.6 means >= 0.3.6, == 0.3.*. So this is expected behaviour. We should reduce the versions as much as possible

And as he mentioned on gitter 
> please do not add more version constraints then necessary. Put the recommended versions in your requirements file (using ==) and give the minimal (and possibly maximal) versions in setup.py.

it may make sense to have 2 different lists for setup.py and requirements.txt as I had mentioned a while back (https://github.com/coala-analyzer/coala/issues/1852#issuecomment-197170604)

**Comment by sebix** [comment]
The comment above summarizes the situation very well: https://github.com/coala-analyzer/coala/issues/1852#issuecomment-197170604

 - put all hard dependencies in setup.py in install_requires using >(=) and <(=) only
   - the software will run with these requirements without errors
   - extensions (bears here) do not necessarily run with these installed (or require the most important ones)
 - put recommended versions for all packages in requirements file using == only
   - this list is longer as it also contains recommended, but not required software

**Comment by wolfv** [comment]
Dear all, 

unfortunately, this breaks on Ubuntu 14.04. The package is also not pip installable anymore.

Not sure if there is an easy fix.
```
Downloading/unpacking coala-bears
  Downloading coala_bears-0.2.1-py3-none-any.whl (14.1MB): 14.1MB downloaded
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2482, in _dep_map
    return self.__dep_map
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2344, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 278, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 1266, in prepare_files
    req_to_install.extras):
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2291, in requires
    dm = self._dep_map
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2484, in _dep_map
    self.__dep_map = self._compute_dependencies()
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2508, in _compute_dependencies
    parsed = next(parse_requirements(distvers))
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2605, in parse_requirements
    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),"version spec")
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2573, in scan_list
    raise ValueError("Expected "+item_name+" in",line,"at",line[p:])
ValueError: ('Expected version spec in', 'coala ~=0.6.0', 'at', ' ~=0.6.0')

Storing debug log for failure in /home/wolfv/.pip/pip.log
```

**Comment by sils1297** [comment]
@wolfv https://github.com/coala-analyzer/coala/wiki/FAQ#installation-is-failing-help should solve your problem.

@AbdealiJK @Makman2 `~=` was a convenience alias, right? Maybe we can avoid the problem for debian users by just writing down the whole thing instead of using `~=` to be backwards compatible?

**Comment by wolfv** [comment]
Hey @sils1297 thanks! good to know there is a workaround. 

However, it will be hard to get everyone on board with this tool if it requires workarounds to install :)
Hopefully you can make it backwards compatible, I guess at work we'll be stuck to Ubuntu 14.04 for a while longer.

**Comment by sils1297** [comment]
@wolfv I'll brew something up right now, then I can give you a command to test it about today. I'm not sure this is possible because it would have to apply for all our dependencies as well. A virtualenv is usually even the recommended way of dealing with things since it avoids problems with other pip installed packages, I can understand that non pythonists or people not accustomed to venvs might find it complicated though.

**Comment by sils1297** [comment]
Related PR: https://gitlab.com/coala/PyPrint/merge_requests/6 needs review, PRs to coala and coala-bears will be following after merge and PyPrint release.

**Comment by sils1297** [comment]
@wolfv can you run `pip install coala --pre` and tell us if that works? I'll update coala-bears in the meantime to be backwards compatible as well.

**Comment by sils1297** [comment]
@wolfv you should now be able to install the prerelease of coala-bears without problems, a confirmation would be nice, maybe something is still using the `~=` operator in some dependency, then we have to carry this upstream.

**Comment by wolfv** [comment]
Hey thanks a lot for the hard work! 

I tried it, but unfortunately there seems to be some other issue with the stars. Ubuntu's pip version seems to be really too old ...

```
pip3 --version
pip 1.5.4 from /usr/lib/python3/dist-packages (python 3.4)
```

```
sudo pip3 install coala --pre
[sudo] password for wolfv: 
Downloading/unpacking coala
  Downloading coala-0.7.0.dev20160525083409-py3-none-any.whl (152kB): 152kB downloaded
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2482, in _dep_map
    return self.__dep_map
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2344, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 278, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 1266, in prepare_files
    req_to_install.extras):
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2291, in requires
    dm = self._dep_map
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2484, in _dep_map
    self.__dep_map = self._compute_dependencies()
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2508, in _compute_dependencies
    parsed = next(parse_requirements(distvers))
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2605, in parse_requirements
    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),"version spec")
  File "/usr/share/python-wheels/setuptools-3.3-py2.py3-none-any.whl/pkg_resources.py", line 2583, in scan_list
    "Expected ',' or end-of-list in",line,"at",line[p:]
ValueError: ("Expected ',' or end-of-list in", 'PyPrint >=0.2.4,==0.2.*', 'at', '*')

Storing debug log for failure in /home/wolfv/.pip/pip.log
```

**Comment by sils1297** [comment]
Argh shit. @wolfv it was worth a try but in this case I think we can't do anything without probably breaking coala for lots of users. :/

**Comment by wolfv** [comment]
no prob! hope we're upgrading to 16.04 soon :)

**Comment by sils1297** [comment]
seems that'll solve it:http://packages.ubuntu.com/xenial/python-pip

**Comment by sils1297** [comment]
What do we do with this? Is it still relevant? Nothing here that a virtualenv or an upgrade wouldn't solve, rght?
